core/rawdb/accessors_batch_event.go

```diff
@@ -0,0 +1,157 @@
+package rawdb
+
+import (
+	"bytes"
+	"math/big"
+	"strconv"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rlp"
+)
+
+// ChunkRange represents the range of blocks within a chunk.
+type ChunkRange struct {
+	StartBlockNumber uint64
+	EndBlockNumber   uint64
+}
+
+// FinalizedBatchMeta holds metadata for finalized batches.
+type FinalizedBatchMeta struct {
+	BatchHash            common.Hash
+	TotalL1MessagePopped uint64
+}
+
+// WriteBatchEventSyncedL1BlockNumber updates the database with the latest synced L1 block number related to batch events in the database.
+func WriteBatchEventSyncedL1BlockNumber(db ethdb.KeyValueWriter, l1BlockNumber uint64) {
+	value := big.NewInt(0).SetUint64(l1BlockNumber).Bytes()
+	if err := db.Put(batchEventSyncedL1BlockNumberKey, value); err != nil {
+		log.Crit("failed to update synced L1 block number for batch event", "err", err)
+	}
+}
+
+// ReadBatchEventSyncedL1BlockNumber fetches the highest synced L1 block number associated with batch events from the database.
+func ReadBatchEventSyncedL1BlockNumber(db ethdb.Reader) *uint64 {
+	data, err := db.Get(batchEventSyncedL1BlockNumberKey)
+	if err != nil && isNotFoundErr(err) {
+		return nil
+	}
+	if err != nil {
+		log.Crit("failed to read synced L1 block number from database", "err", err)
+	}
+
+	number := new(big.Int).SetBytes(data)
+	if !number.IsUint64() {
+		log.Crit("unexpected synced L1 block number in database", "number", number)
+	}
+
+	batchEventSyncedL1BlockNumber := number.Uint64()
+	return &batchEventSyncedL1BlockNumber
+}
+
+// WriteBatchChunkRanges writes the block ranges for each chunk within a batch to the database.
+// It serializes the chunk ranges using RLP and stores them under a key derived from the batch index.
+func WriteBatchChunkRanges(db ethdb.KeyValueWriter, batchIndex uint64, chunkRanges []*ChunkRange) {
```

Comment:
What's the rationale for storing batches by index as opposed to batch hash? (Seems okay to me, just curious about the considerations behind this decision.)

Reply:
because the service listens to L1 rollup events and updates db like: commit -> commit -> revert -> ...
both `batchIndex` and `batchHash` lgtm.